src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -450,80 +454,225 @@ private bool TryCreateCountSegment(string segmentText)
+
+            return true;
+        }
 
-                keySegmentsForPreviousPathSegment.Push(key);
-                this.parsedSegments.Remove(key);
+        /// <summary>
+        /// Try to handle the segment as $each.
+        /// </summary>
+        /// <param name="identifier">The identifier that was parsed from this raw segment.</param>
+        /// <param name="parenthesisExpression">The query portion was parsed from this raw segment.
+        /// This value can be null if there is no query portion.</param>
+        /// <returns>Whether the segment was $each.</returns>
+        private bool TryCreateEachSegment(string identifier, string parenthesisExpression)
+        {
+            if (!IdentifierIs(UriQueryConstants.EachSegment, identifier))
+            {
+                return false;
             }
 
-            // Get the previous Path segment
-            ODataPathSegment previous = this.parsedSegments[this.parsedSegments.Count - 1];
-            NavigationPropertySegment navPropSegment = previous as NavigationPropertySegment;
+            // $each is not supposed to have parenthesis expressions after it.
+            if (parenthesisExpression != null)
+            {
+                throw ExceptionUtil.CreateSyntaxError();
+            }
 
-            // $ref can not be applied for entity set. It can be applied only on entity and collection of entities
-            if ((navPropSegment == null) || navPropSegment.TargetKind != RequestTargetKind.Resource)
+            ODataPathSegment prevSegment = this.parsedSegments.Last();
+            if (lastNavigationSource == null)
             {
-                throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.PathParser_EntityReferenceNotSupported(previous.Identifier));
+                throw new ODataException(ODataErrorStrings.RequestUriProcessor_NoNavigationSourceFound(UriQueryConstants.EachSegment));
             }
 
-            // Since this $ref segment, remove the previous segment from parsedSegments as we need to repalce that with new ref segment
-            this.parsedSegments.Remove(previous);
+            if (lastNavigationSource is IEdmSingleton || prevSegment is KeySegment)
+            {
+                throw new ODataException(ODataErrorStrings.RequestUriProcessor_CannotApplyEachOnSingleEntities(lastNavigationSource.Name));
+            }
 
-            // If this is a navigation property, find target navigation source
-            IEdmPathExpression bindingPath;
-            var targetNavigationSource = this.parsedSegments[parsedSegments.Count - 1].TargetEdmNavigationSource.FindNavigationTarget(navPropSegment.NavigationProperty, BindingPathHelper.MatchBindingPath, this.parsedSegments, out bindingPath);
+            EachSegment eachSegment = new EachSegment(
+                lastNavigationSource,
+                prevSegment is TypeSegment || prevSegment is FilterSegment ? prevSegment.TargetEdmType : null);

```

Comment:
Can you just do EachSegment eachSegment = new EachSegment(lastNavigationSource,prevSegment.TargetEdmType)?

Reply:
Stepped through the debugger a several times. Just posting results here (follow-up response after):

1. If previous segment was TypeSegment (e.g. "People/Fully.Qualified.Namespace.Employee/$each")

prevSegment.TargetEdmType:
TargetEdmType	{Fully.Qualified.Namespace.Employee}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.Csdl.CsdlSemantics.CsdlSemanticsEntityTypeDefinition}

lastNavigationSource.EntityType():
TargetEdmType	{Fully.Qualified.Namespace.Person}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.Csdl.CsdlSemantics.CsdlSemanticsEntityTypeDefinition}

2. If previous segment was FilterSegment (e.g. "People/$filter(@p1)/$each")

prevSegment.TargetEdmType:
TargetEdmType	{Fully.Qualified.Namespace.Person}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.Csdl.CsdlSemantics.CsdlSemanticsEntityTypeDefinition}

lastNavigationSource.EntityType():
TargetEdmType	{Fully.Qualified.Namespace.Person}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.Csdl.CsdlSemantics.CsdlSemanticsEntityTypeDefinition}

3. If previous segment was FilterSegment whose previous segment was TypeSegment (e.g. "People/Fully.Qualified.Namespace.Employee/$filter(@p1)/$each")

prevSegment.TargetEdmType:
TargetEdmType	{Fully.Qualified.Namespace.Person}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.Csdl.CsdlSemantics.CsdlSemanticsEntityTypeDefinition}

lastNavigationSource.EntityType():
TargetEdmType	{Fully.Qualified.Namespace.Person}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.Csdl.CsdlSemantics.CsdlSemanticsEntityTypeDefinition}

4. If previous segment was EntitySetSegment (e.g. "People/$each")

prevSegment.TargetEdmType:
TargetEdmType	{Fully.Qualified.Namespace.Person}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.Csdl.CsdlSemantics.CsdlSemanticsEntityTypeDefinition}

lastNavigationSource.EntityType():
TargetEdmType	{Fully.Qualified.Namespace.Person}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.Csdl.CsdlSemantics.CsdlSemanticsEntityTypeDefinition}

5. If previous segment was NavigationPropertySegment (e.g. "People/1/MyFriendsDogs/$each")

prevSegment.TargetEdmType:
TargetEdmType	{Collection([Fully.Qualified.Namespace.Dog Nullable=False])}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.EdmCollectionType}

lastNavigationSource.EntityType():
TargetEdmType	{Fully.Qualified.Namespace.Dog}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.Csdl.CsdlSemantics.CsdlSemanticsEntityTypeDefinition}

Reply:
From case 3, we need to specifically use the prevSegment.TargetEdmType for both FilterSegment and TypeSegment scenarios to retain the derived type information. Therefore we can't just pass the lastNavigationSource.EntityType().

From case 5, we can't just pass the prevSegment.TargetEdmType as it refers to a collection.

Therefore, I think the code there might be the cleanest form to account for all the cases.

Reply:
How about prevSegment.TargetEdmType.AsElementType()?

---
In reply to: [225761341](https://github.com/OData/odata.net/pull/1110#discussion_r225761341) [](ancestors = 225761341)

Reply:
That works! That also makes sense that there's a property in there somewhere that refers to the singular type.